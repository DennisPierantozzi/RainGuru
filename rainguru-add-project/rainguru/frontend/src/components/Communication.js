import React from "react";
import Popup from "react-popup";
import Map from "./Map/Map";

export default class Communication {
    static useLatestData = true;
    static observed = false;
    static dataCollected = false;
    static imageUrls = [];
    static latestTimestamp = 0;
    static requestTimestamp = [];
    static dataTimestamp = [];

    // getters
    /**
     * Get the boolean representing if we use the latest data.
     */
    static getUseLatestData() {
        return this.useLatestData;
    }

    /**
     * Get the boolean representing if we show observed or predicted data.
     */
    static getObserved() {
        return this.observed;
    }

    /**
     * Get the boolean representing if all the data has been loaded.
     */
    static getDataCollected() {
        return this.dataCollected;
    }

    /**
     * Get the urls for the images with the predicted precipitation.
     */
    static getImageUrls() {
        return this.imageUrls;
    }

    /**
     * Get the timestamp of the first prediction of the latest data.
     */
    static getLatestTimestamp() {
        return this.latestTimestamp;
    }

    /**
     * Get the timestamp that is used when making requests.
     */
    static getRequestTimestamp() {
        return this.requestTimestamp;
    }

    /**
     * Get the timestamp of the first prediction of the data in use.
     */
    static getDataTimestamp() {
        return this.dataTimestamp;
    }

    // setters
    /**
     * Set the boolean representing to if we use the latest data.
     */
    static setUseLatestData(latest) {
        this.useLatestData = latest;
    }

    /**
     * Set the boolean representing if we show observed or predicted data.
     */
    static setObserved(observed) {
        this.observed = observed;
    }

    /**
     * Set the timestamp that the requests will use
     */
    static setRequestTimestamp(timestamp) {
        this.requestTimestamp = timestamp;
    }


    /**
     * Fetches new data from the back-end.
     */
    static async fetchUrls() {
        let url = "api/fetch?observed=" + this.observed;
        // only add timestamp if not fetching latest data
        if (!this.useLatestData) {
            url += "&timestamp=" + this.requestTimestamp;
        }
        // make request to retrieve predictions
        await fetch(url)
          .then(response => {
            if (response.status > 400) {
              throw new Error();
            }
            return response.json();
          })
          .then(responseJson => {
            // store predictions and create images for them
            this.imageUrls = responseJson.urls;
            this.dataTimestamp = new Date(responseJson.timestamp * 1000);
            if(responseJson.exception_active) {
                Popup.alert("Message from server:\n" + responseJson.exception_message, "Warning: Model might be using old data.");
                console.log(responseJson.exception_message);
            }
            Map.preloadedImages();
            this.dataCollected = true;
          });
    }

    /**
     * Fetches the timestamp of the first prediction from the latest data.
     *
     * @returns {Promise<boolean>} representing if there is new data available.
     */
    static async checkNewData() {
        let dataUpdated = false;
        // make a request for the latest data first prediction timestamp
        await fetch("api/check")
            .then(response => {
                if (response.status > 400) {
                  throw new Error();
                }
                return response.json();
            })
            .then(responseJson => {
                // check if there is new data via the timestamp received
                const newTimestamp = responseJson.timestamp;
                if (newTimestamp !== this.latestTimestamp) {
                    this.latestTimestamp = newTimestamp;
                    dataUpdated = true;
                }
            });
        return dataUpdated;
    }

    /**
     * Fetches the data about a selected location for a selected 100 minutes.
     *
     * @param x the x coordinate of the 480x480 images
     * @param y the y coordinate of the 480x480 images
     * @returns {Promise<any>} the precipitation of the location for a selected 100 minutes.
     */
    static async fetchPrecipitation(x, y) {
        let url = "api/precipitation?x=" + x + "&y=" + y + "&observed=" + this.observed;
        // only add timestamp if not fetching latest data
        if (!this.useLatestData) {
            url += "&timestamp=" + this.requestTimestamp;
        }

        // make a request for the precipitation information of the give location
        return await fetch(url)
            .then(response => {
                if (response.status > 400) {
                    throw new Error();
                }
                return response.json();
            })
            .then(responseJson => {
                // return the fetched precipitation
                return responseJson.precipitation;
            })
    }

    /**
     * Fetch the times of what data is all available
     *
     * @returns {Promise<{available: *, predictedTimes: *, observedTimes: *}>} the availability of data times
     */
    static async fetchDataAvailability() {
        // make a request to get the availability times
        return await fetch("api/available")
            .then(response => {
                if (response.status > 400) {
                    throw new Error()
                }
                return response.json();
            })
            .then(responseJson => {
                // return a boolean representing if data is being stored and the available data times
                return {
                    available: responseJson.store_data,
                    predictedTimes: responseJson.predictions_stored,
                    observedTimes: responseJson.observations_stored
                };
            })
    }
}