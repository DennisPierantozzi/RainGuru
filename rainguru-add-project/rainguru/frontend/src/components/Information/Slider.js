import React from "react";
import Communication from "../Communication"
import AnimationBar from "./AnimationBar";

export default class Slider {

    // stores the latest predictions, slider state (0 to 19), and latitude and longitude of the selected location
    static predictions;
    static sliderValue;
    static lastLat;
    static lastLong;

    /**
     * Retrieves and displays the precipitation data at a specific location.
     * @param latitude the latitude of the specified location.
     * @param longitude the longitude of the specified location.
     */
    static async showPrecipitationData(latitude, longitude) {
        try {
            // update the last lat long value
            Slider.lastLat = latitude;
            Slider.lastLong = longitude;

            // calculate the corresponding rain value stored in the data matrix for the given latitude and longitude
            const y_temp = (latitude - 50.21) / (54.71 - 50.21);
            const x_temp = (longitude - 1.90) / (9.30 - 1.90);
            const y = Math.round(y_temp * 700);
            const x = Math.round(x_temp * 700);
            const precipitation = await Communication.fetchPrecipitation(699 - y, x);

            // find the maximum precipitation amount for the selected location
            const predictions = [];
            let maxRain = 0;
            for (let t = 0; t < 20; t++) {
                if (precipitation[t] > maxRain) {
                    maxRain = precipitation[t];
                }
                predictions.push(precipitation[t]);
            }

            // update the static predictions variable
            Slider.predictions = predictions;

            // dynamically update the interval information and append newly created node to "intervalInfo" div
            const intervalInfoDiv = document.getElementById("intervalInfo");
            intervalInfoDiv.innerHTML = "";
            let timeString = AnimationBar.fullMarks[Slider.sliderValue];
            let intervalInfoDivNode;
            // when precipitation amount is a whole number, add .0 to it
            if (Slider.predictions[Slider.sliderValue] % 1 === 0) {
                intervalInfoDivNode = document.createTextNode(timeString + " - " + Slider.predictions[Slider.sliderValue] + ".0 mm/h");
            } else {
                intervalInfoDivNode = document.createTextNode(timeString + " - " + Slider.predictions[Slider.sliderValue] + " mm/h");
            }
            intervalInfoDiv.appendChild(intervalInfoDivNode);

            // dynamically change maximum rain amount for the given location
            const maxRainDiv = document.getElementById("maxRain");
            if (maxRain % 1 === 0) {
                maxRainDiv.innerHTML = "Max: " + maxRain + ".0 mm/h";
            } else {
                maxRainDiv.innerHTML = "Max: " + maxRain + " mm/h";
            }

            // dynamically set maximum scale value
            const maxScaleDiv = document.getElementById("max-scale");
            let maxScale = maxRain.toFixed(2);
            // when maxRain is less than 0.1, set the maxScale to be 0.1
            if (maxRain <= 0.1) {
                maxScale = 0.1;
            }
            maxScaleDiv.innerHTML = maxScale;

            // dynamically set middle scale value
            const midScaleDiv = document.getElementById("mid-scale");
            let midScaleValue;
            // when maxScale is less than 0.1, set the midScaleValue to be 0.1
            if (maxRain <= 0.1) {
                midScaleValue = 0.05;
            } else {
                midScaleValue = (maxRain / 2).toFixed(2);
            }
            midScaleDiv.innerHTML = midScaleValue;

            // call a helper method to create a precipitation chart
            Slider.makeChart(predictions, maxScale);
        } catch (e) {
            console.log(e);
        }
    }

    /**
     * Creates a precipitation chart (filled line chart) and append it to "chart" div.
     * @param predictions Predictions for the next 100 minutes that need to be graphed
     * @param maxScale Indication of the maximum y-value
     */
    static makeChart(predictions, maxScale) {
        const currentTime = new Date();
        const timeStamps = [];
        for (let i = 0; i < 20; i++) {
            timeStamps.push(i);
        }
        const chart = document.getElementById("chart").getContext("2d");
        const predictionGraph = new Chart(chart, {
            type: "line",
            data: {
                labels: timeStamps,
                datasets: [{
                    fill: true,
                    data: predictions,
                    borderColor: "#1167b1",
                    backgroundColor: "rgba(173, 216, 230, 0.7)"
                }]
            },
            options: {
                maintainAspectRatio: false,
                scales: {
                    xAxes: [{
                        display: false,
                        gridLines: {
                            color: "rgba(0, 0, 0, 0)"
                        }
                    }],
                    yAxes: [{
                        display: false,
                        ticks: {
                            beginAtZero: true,
                            max: maxScale * 1.05
                        },
                        gridLines: {
                            color: "rgba(0, 0, 0, 0)",
                        },
                    }]
                },
                legend:
                    {
                        display: false
                    }
            }
        })
    }
}
