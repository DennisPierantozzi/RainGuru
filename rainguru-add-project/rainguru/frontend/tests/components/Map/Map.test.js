import React from "react";
import {render, screen} from "@testing-library/react";
import L from "leaflet";
import Popup from "react-popup";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import Map from "../../../src/components/Map/Map";
import Communication from "../../../src/components/Communication";
import Slider from "../../../src/components/Information/Slider";

const imageUrlsUsed = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
        "10", "11", "12", "13", "14", "15", "16", "17", "18", "19"]

beforeEach(() => {
    // set up communication with fake image urls
    Communication.imageUrls = imageUrlsUsed;

    // mock the search functionality since it gives problems when testing
    Map.addMapSearch = jest.fn();
});
afterEach(() => {
    jest.clearAllMocks();
});

// the function coverage for Map.js is quite low, this is since the map events call a method for their code
// this method call is hard to test since the event is quite inaccessible, this costs a lot of function calls
// the rest of the missing coverage is since the testing environment can't load the geosearch part of the map
describe("Map tests", () => {
    it("without search functionality matches snapshot", () => {
        // render the map without the search functionality and check it with a snapshot
        render(<div id="app" data-testid="app">
            <div id="map"></div>
            <div id="locationField"></div>
            </div>);
        Map.render();
        const state = renderer.create(screen.getByTestId("app").innerHTML).toJSON();
        expect(state).toMatchSnapshot();
    });

    it("generates image layers correctly", () => {
        // state before the method is called
        Map.heatLayers = [];
        Map.nextLayers = [];
        expect(Map.heatLayers).toStrictEqual([]);
        expect(Map.nextLayers).toStrictEqual([]);

        // generate the result list
        let heatLayersResult = [];
        let nextLayersResult = [];
        const imageBoundsTest = [[54.71, 1.90], [50.21, 9.30]];
        for (let i = 0; i < 20; i++) {
            heatLayersResult.push(L.imageOverlay(imageUrlsUsed[i], imageBoundsTest, {opacity: 0.4}));
            if (i !== 0) {
                nextLayersResult.push(L.imageOverlay(imageUrlsUsed[i], imageBoundsTest, {opacity: 0}));
            }
        }
        nextLayersResult.push(L.imageOverlay(imageUrlsUsed[0], imageBoundsTest, {opacity: 0}));

        // call the method and check the made lists
        Map.setHeatLayers();
        expect(Map.heatLayers).toStrictEqual(heatLayersResult);
        expect(Map.nextLayers).toStrictEqual(nextLayersResult);
    });

    it("set the active heat layer", () => {
        // state before the method is called
        Map.heatLayers = [];
        Map.nextLayers = [];
        Map.currentLayer = new L.LayerGroup();
        expect(Map.heatLayers).toStrictEqual([]);
        expect(Map.nextLayers).toStrictEqual([]);
        expect(Map.currentLayer).toStrictEqual(new L.LayerGroup());

        // set the layers that need to be used, these are tested in a different test
        Map.setHeatLayers();

        // generate the shown layer to check the result
        let shownLayer1 = new L.LayerGroup();
        shownLayer1.addLayer(Map.heatLayers[6]);
        shownLayer1.addLayer(Map.nextLayers[6]);
        let shownLayer2 = new L.LayerGroup();
        shownLayer2.addLayer(Map.heatLayers[17]);
        shownLayer2.addLayer(Map.nextLayers[17]);

        // call the method and check the loaded layers
        Map.activateHeatLayer(6);
        expect(Map.currentLayer).toStrictEqual(shownLayer1);

        // call the method and check the loaded layers while there are other layers loaded
        Map.activateHeatLayer(17);
        expect(Map.currentLayer).toStrictEqual(shownLayer2);
    });

    describe("map.on event methods tests", () => {
        describe("location out of bounds", () => {
            it("latitude greater than max", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check if the alert was given correctly
                Map.markerHandling({latlng: {lat: 55, lng: 5.6}}, false, false);
                expect(alertMethod).toBeCalledTimes(1);
                expect(alertMethod).toBeCalledWith("Your selected location is outside the bounds of the precipitation map", "Location out of bounds");
                expect(showPrecipitationMethod).toBeCalledTimes(0);
                expect(spy1).toBeCalledTimes(0);
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(0);
                expect(spy4).toBeCalledTimes(0);
            });

            it("latitude lesser than max", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check if the alert was given correctly
                Map.markerHandling({latlng: {lat: 50, lng: 5.6}}, false, false);
                expect(alertMethod).toBeCalledTimes(1);
                expect(alertMethod).toBeCalledWith("Your selected location is outside the bounds of the precipitation map", "Location out of bounds");
                expect(showPrecipitationMethod).toBeCalledTimes(0);
                expect(spy1).toBeCalledTimes(0);
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(0);
                expect(spy4).toBeCalledTimes(0);
            });

            it("longitude lesser than max", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check if the alert was given correctly
                Map.markerHandling({latlng: {lat: 52.46, lng: 1}}, false, false);
                expect(alertMethod).toBeCalledTimes(1);
                expect(alertMethod).toBeCalledWith("Your selected location is outside the bounds of the precipitation map", "Location out of bounds");
                expect(showPrecipitationMethod).toBeCalledTimes(0);
                expect(spy1).toBeCalledTimes(0);
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(0);
                expect(spy4).toBeCalledTimes(0);
            });

            it("longitude greater than max", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check if the alert was given correctly
                Map.markerHandling({latlng: {lat: 52.46, lng: 10}}, false, false);
                expect(alertMethod).toBeCalledTimes(1);
                expect(alertMethod).toBeCalledWith("Your selected location is outside the bounds of the precipitation map", "Location out of bounds");
                expect(showPrecipitationMethod).toBeCalledTimes(0);
                expect(spy1).toBeCalledTimes(0);
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(0);
                expect(spy4).toBeCalledTimes(0);
            });

            it("out of bounds of search, requiring fly to back when no marker was already present", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check if the alert was given correctly, no marker was present already
                Map.markerHandling({location: {y: 55, x: 5.6}}, true, false);
                expect(alertMethod).toBeCalledTimes(1);
                expect(alertMethod).toBeCalledWith("Your selected location is outside the bounds of the precipitation map", "Location out of bounds");
                expect(showPrecipitationMethod).toBeCalledTimes(0);
                expect(spy1).toBeCalledTimes(1);
                expect(spy1).toBeCalledWith(L.latLng(52.45, 5.415), 8);
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(0);
                expect(spy4).toBeCalledTimes(0);
            });

            it("out of bounds of search, requiring fly to back when a marker was already present", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // set up the marker to already be present on the map
                Map.marker = L.marker([52, 6]).addTo(Map.map);

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check if the alert was given correctly, a marker was present already
                Map.markerHandling({location: {y: 55, x: 5.6}}, true, false);
                expect(alertMethod).toBeCalledTimes(1);
                expect(alertMethod).toBeCalledWith("Your selected location is outside the bounds of the precipitation map", "Location out of bounds");
                expect(showPrecipitationMethod).toBeCalledTimes(0);
                expect(spy1).toBeCalledTimes(1);
                expect(spy1).toBeCalledWith(L.latLng(52, 6), 8);
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(0);
                expect(spy4).toBeCalledTimes(0);
            });
        });

        describe("location in bounds", () => {
            it("place marker in bounds when no marker was already present, event was called with search", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // clear any possibly available marker and set a present marker
                const setupMarker = L.marker([52.1, 6.1]);
                Map.marker = setupMarker.addTo(Map.map);

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check the methods are called correctly
                Map.markerHandling({location: {y: 53, x: 6.4}}, true, false);
                expect(alertMethod).toBeCalledTimes(0);
                expect(showPrecipitationMethod).toBeCalledTimes(1);
                expect(spy1).toBeCalledTimes(1); // setView is called once when setZoom is called
                expect(spy2).toBeCalledTimes(1);
                expect(spy2).toBeCalledWith(setupMarker);
                expect(spy3).toBeCalledTimes(1);
                expect(spy3).toBeCalledWith(8);
                expect(spy4).toBeCalledTimes(0);
            });

            it("place marker in bounds when a marker was already present, event was called with search", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // clear any possibly available marker
                Map.marker = undefined;

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check the methods are called correctly
                Map.markerHandling({location: {y: 53, x: 6.4}}, true, false);
                expect(alertMethod).toBeCalledTimes(0);
                expect(showPrecipitationMethod).toBeCalledTimes(1);
                expect(spy1).toBeCalledTimes(1); // setView is called once when setZoom is called
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(1);
                expect(spy3).toBeCalledWith(8);
                expect(spy4).toBeCalledTimes(0);
            });

            it("place marker in bounds when no marker was already present, event was called with found and uses min zoom", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // clear any possibly available marker
                Map.marker = undefined;
                // set the zoom of the map to that the flyTo method is called correctly
                Map.map.setZoom(7);
                expect(Map.map.getZoom()).toBe(7);

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check the methods are called correctly
                Map.markerHandling({latlng: {lat: 53, lng: 6.4}}, false, true);
                expect(alertMethod).toBeCalledTimes(0);
                expect(showPrecipitationMethod).toBeCalledTimes(1);
                expect(spy1).toBeCalledTimes(1); // setView is called once when flyTo is called
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(0);
                expect(spy4).toBeCalledTimes(1);
                expect(spy4).toBeCalledWith(L.latLng(53, 6.4), 8);
            });

            it("place marker in bounds when no marker was already present, event was called with found and uses map zoom", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // clear any possibly available marker
                Map.marker = undefined;
                // set the zoom of the map to that the flyTo method is called correctly
                Map.map.setZoom(12);
                expect(Map.map.getZoom()).toBe(12);

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check the methods are called correctly
                Map.markerHandling({latlng: {lat: 53, lng: 6.4}}, false, true);
                expect(alertMethod).toBeCalledTimes(0);
                expect(showPrecipitationMethod).toBeCalledTimes(1);
                expect(spy1).toBeCalledTimes(1); // setView is called once when flyTo is called
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(0);
                expect(spy4).toBeCalledTimes(1);
                expect(spy4).toBeCalledWith(L.latLng(53, 6.4), 12);
            });

            it("place marker in bounds when no marker was already present, event was called with click", () => {
                // render the map, this is tested in other tests
                render(<div id="app" data-testid="app">
                    <div id="map"></div>
                    <div id="locationField"></div>
                </div>);
                Map.render();

                // clear any possibly available marker
                Map.marker = undefined;
                // set the zoom of the map to that the flyTo method is called correctly
                Map.map.setZoom(7);
                expect(Map.map.getZoom()).toBe(7);

                // create mocks and spies
                const alertMethod = jest.fn();
                Popup.alert = alertMethod;
                const showPrecipitationMethod = jest.fn();
                Slider.showPrecipitationData = showPrecipitationMethod;
                const spy1 = jest.spyOn(Map.map, "setView");
                const spy2 = jest.spyOn(Map.map, "removeLayer");
                const spy3 = jest.spyOn(Map.map, "setZoom");
                const spy4 = jest.spyOn(Map.map, "flyTo");

                // call the method and check the methods are called correctly
                Map.markerHandling({latlng: {lat: 53, lng: 6.4}}, false, false);
                expect(alertMethod).toBeCalledTimes(0);
                expect(showPrecipitationMethod).toBeCalledTimes(1);
                expect(spy1).toBeCalledTimes(1); // setView is called once when flyTo is called
                expect(spy2).toBeCalledTimes(0);
                expect(spy3).toBeCalledTimes(0);
                expect(spy4).toBeCalledTimes(1);
                expect(spy4).toBeCalledWith(L.latLng(53, 6.4), 7);
            });
        });
    });
});
