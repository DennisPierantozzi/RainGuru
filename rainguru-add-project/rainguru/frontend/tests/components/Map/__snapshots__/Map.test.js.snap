// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map tests without search functionality matches snapshot 1`] = `"<div id=\\"map\\" class=\\"leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom\\" style=\\"position: relative;\\" tabindex=\\"0\\"><div class=\\"leaflet-pane leaflet-map-pane\\" style=\\"left: 0px; top: 0px;\\"><div class=\\"leaflet-pane leaflet-tile-pane\\"><div class=\\"leaflet-layer \\" style=\\"z-index: 1;\\"><div class=\\"leaflet-tile-container leaflet-zoom-animated\\" style=\\"z-index: 18; left: 0px; top: 0px;\\"><img alt=\\"\\" role=\\"presentation\\" src=\\"https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer/tile/7/42/65\\" class=\\"leaflet-tile\\" style=\\"width: 256px; height: 256px; left: -237px; top: -5px;\\"></div></div></div><div class=\\"leaflet-pane leaflet-overlay-pane\\"><img class=\\"leaflet-image-layer\\" style=\\"z-index: 1; opacity: 0.4; left: -320px; top: -347px; width: 674px; height: 673px;\\" src=\\"0\\" alt=\\"\\"><img class=\\"leaflet-image-layer\\" style=\\"z-index: 1; opacity: 0; left: -320px; top: -347px; width: 674px; height: 673px;\\" src=\\"1\\" alt=\\"\\"></div><div class=\\"leaflet-pane leaflet-shadow-pane\\"></div><div class=\\"leaflet-pane leaflet-marker-pane\\"></div><div class=\\"leaflet-pane leaflet-tooltip-pane\\"></div><div class=\\"leaflet-pane leaflet-popup-pane\\"></div></div><div class=\\"leaflet-control-container\\"><div class=\\"leaflet-top leaflet-left\\"><div class=\\"leaflet-control-zoom leaflet-bar leaflet-control\\"><a class=\\"leaflet-control-zoom-in\\" href=\\"#\\" title=\\"Zoom in\\" role=\\"button\\" aria-label=\\"Zoom in\\" aria-disabled=\\"false\\"><span aria-hidden=\\"true\\">+</span></a><a class=\\"leaflet-control-zoom-out leaflet-disabled\\" href=\\"#\\" title=\\"Zoom out\\" role=\\"button\\" aria-label=\\"Zoom out\\" aria-disabled=\\"true\\"><span aria-hidden=\\"true\\">−</span></a></div></div><div class=\\"leaflet-top leaflet-right\\"></div><div class=\\"leaflet-bottom leaflet-left\\"><div class=\\"leaflet-control-scale leaflet-control\\"><div class=\\"leaflet-control-scale-line\\" style=\\"width: 67px;\\">50 km</div></div></div><div class=\\"leaflet-bottom leaflet-right\\"><div class=\\"leaflet-control-attribution leaflet-control\\"><a href=\\"https://leafletjs.com\\" title=\\"A JavaScript library for interactive maps\\">Leaflet</a> <span aria-hidden=\\"true\\">|</span> © <a href=\\"https://www.esri.com/en-us/home\\">Esri</a></div></div></div></div><div id=\\"locationField\\"></div>"`;
